rust/worker/src/segment/metadata_segment.rs

```diff
@@ -0,0 +1,1158 @@
+
+        Ok(MetadataSegmentWriter {
+            full_text_index_writer: Some(full_text_index_writer),
+            string_metadata_index_writer: Some(string_metadata_index_writer),
+            bool_metadata_index_writer: Some(bool_metadata_index_writer),
+            f32_metadata_index_writer: Some(f32_metadata_index_writer),
+            u32_metadata_index_writer: Some(u32_metadata_index_writer),
+        })
+    }
+
+    pub async fn write_to_blockfiles(&mut self) -> Result<(), MetadataSegmentError> {
+        let mut full_text_index_writer = self
+            .full_text_index_writer
+            .take()
+            .ok_or_else(|| MetadataSegmentError::NoWriter)?;
+        let res = full_text_index_writer.write_to_blockfiles().await;
+        self.full_text_index_writer = Some(full_text_index_writer);
+        match res {
+            Ok(_) => {}
+            Err(_) => return Err(MetadataSegmentError::BlockfileWriteError),
+        }
+
+        let mut string_metadata_index_writer = self
+            .string_metadata_index_writer
+            .take()
+            .ok_or_else(|| MetadataSegmentError::NoWriter)?;
+        let res = string_metadata_index_writer.write_to_blockfile().await;
+        self.string_metadata_index_writer = Some(string_metadata_index_writer);
+        match res {
+            Ok(_) => {}
+            Err(_) => return Err(MetadataSegmentError::BlockfileWriteError),
+        }
+
+        let mut bool_metadata_index_writer = self
+            .bool_metadata_index_writer
+            .take()
+            .ok_or_else(|| MetadataSegmentError::NoWriter)?;
+        let res = bool_metadata_index_writer.write_to_blockfile().await;
+        self.bool_metadata_index_writer = Some(bool_metadata_index_writer);
+        match res {
+            Ok(_) => {}
+            Err(_) => return Err(MetadataSegmentError::BlockfileWriteError),
+        }
+
+        let mut f32_metadata_index_writer = self
+            .f32_metadata_index_writer
+            .take()
+            .ok_or_else(|| MetadataSegmentError::NoWriter)?;
+        let res = f32_metadata_index_writer.write_to_blockfile().await;
+        self.f32_metadata_index_writer = Some(f32_metadata_index_writer);
+        match res {
+            Ok(_) => {}
+            Err(_) => return Err(MetadataSegmentError::BlockfileWriteError),
+        }
+
+        let mut u32_metadata_index_writer = self
+            .u32_metadata_index_writer
+            .take()
+            .ok_or_else(|| MetadataSegmentError::NoWriter)?;
+        let res = u32_metadata_index_writer.write_to_blockfile().await;
+        self.u32_metadata_index_writer = Some(u32_metadata_index_writer);
+        match res {
+            Ok(_) => {}
+            Err(_) => return Err(MetadataSegmentError::BlockfileWriteError),
+        }
+
+        Ok(())
+    }
+}
+
+impl SegmentWriter for MetadataSegmentWriter {
+    fn apply_materialized_log_chunk(
```

Comment:
note to self: This needs to be audited again after materializedLogRecordV2 is merged. Also, TBD add support for updates/deletes/upserts here